libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp

```diff
@@ -122,26 +363,291 @@ namespace hpx { namespace parallel { inline namespace v1 {
+                    hpx::traits::range_iterator_t<Rng>,
+                    typename hpx::traits::range_sentinel<Rng>::type>>()
+                    .call2(std::forward<ExPolicy>(policy), is_seq(),
+                        hpx::util::begin(rng), middle, hpx::util::end(rng)));
+        }
+    } rotate{};
+
+    ///////////////////////////////////////////////////////////////////////////
+    // DPO for hpx::ranges::rotate_copy
+    template <typename I, typename O>
+    using rotate_copy_result = hpx::parallel::util::in_out_result<I, O>;
+
+    HPX_INLINE_CONSTEXPR_VARIABLE struct rotate_copy_t final
+      : hpx::functional::tag_fallback<rotate_copy_t>
+    {
+    private:
+        // clang-format off
+        template <typename FwdIter, typename Sent, typename OutIter,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator_v<FwdIter> &&
+                hpx::traits::is_sentinel_for<Sent, FwdIter>::value &&
+                hpx::traits::is_iterator_v<OutIter>
+            )>
+        // clang-format on
+        friend rotate_copy_result<FwdIter, OutIter> tag_fallback_dispatch(
+            hpx::ranges::rotate_copy_t, FwdIter first, FwdIter middle,
+            Sent last, OutIter dest_first)
+        {
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter>),
+                "Requires at least forward iterator.");
+            static_assert((hpx::traits::is_output_iterator_v<OutIter>),
+                "Requires at least output iterator.");
+
+            return hpx::parallel::v1::detail::rotate_copy<
+                rotate_copy_result<FwdIter, OutIter>>()
+                .call(hpx::execution::seq, first, middle, last, dest_first);
+        }
+
+        // clang-format off
+        template <typename ExPolicy, typename FwdIter1, typename Sent,
+            typename FwdIter2, HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator_v<FwdIter1> &&
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_sentinel_for<Sent, FwdIter1>::value &&
+                hpx::traits::is_iterator_v<FwdIter2>
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            rotate_copy_result<FwdIter1, FwdIter2>>::type
+        tag_fallback_dispatch(hpx::ranges::rotate_copy_t, ExPolicy&& policy,
+            FwdIter1 first, FwdIter1 middle, Sent last, FwdIter2 dest_first)
+        {
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter1>),
+                "Requires at least forward iterator.");
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter2>),
+                "Requires at least forward iterator.");
+
+            using is_seq = std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
+                    !hpx::traits::is_forward_iterator_v<FwdIter1>>;
```

Comment:
I don't think this is correct. A few lines above you verify that the iterator is a `forward_iterator`. I think the correct way would be:
```
        using is_seq = std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_bidirectional_iterator_v<FwdIter1>>;
```

Reply:
Sorry for that, will amend it.